File: src/main/java/com/github/frcsty/frozenactions/time/TimeAPI.java
package com.github.frcsty.frozenactions.time;

import org.jetbrains.annotations.NotNull;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;
import java.util.concurrent.TimeUnit;

public final class TimeAPI {

    private static final int DAYS_IN_WEEK = 7;
    private static final int DAYS_IN_MONTH = 30;
    private static final int DAYS_IN_YEAR = 365;

    public @NotNull TimeResult parseTime(final @NotNull String input) {
        long seconds = 0L;
        final TimeScanner scanner = new TimeScanner(
                input
                        .replace(" ", "")
                        .replace("and", "")
                        .replace(",", "")
                        .toLowerCase()
        );

        long next;
        while (scanner.hasNext()) {
            next = scanner.nextLong();
            switch (TimeType.getType(scanner.nextString())) {
                case SECOND -> seconds += next;
                case MINUTE -> seconds += TimeUnit.MINUTES.toSeconds(next);
                case HOUR -> seconds += TimeUnit.HOURS.toSeconds(next);
                case DAY -> seconds += TimeUnit.DAYS.toSeconds(next);
                case WEEK -> seconds += TimeUnit.DAYS.toSeconds(next * DAYS_IN_WEEK);
                case MONTH -> seconds += TimeUnit.DAYS.toSeconds(next * DAYS_IN_MONTH);
                case YEAR -> seconds += TimeUnit.DAYS.toSeconds(next * DAYS_IN_YEAR);
            }
        }

        return new TimeResult(seconds);
    }

    private enum TimeType {

        SECOND("s", "sec", "secs", "second", "seconds"),
        MINUTE("m", "min", "mins", "minute", "minutes"),
        HOUR("h", "hr", "hrs", "hour", "hours"),
        DAY("d", "dy", "dys", "day", "days"),
        WEEK("w", "week", "weeks"),
        MONTH("mo", "mon", "mnth", "month", "months"),
        YEAR("y", "yr", "yrs", "year", "years");

        final Set<String> matches;

        TimeType(final String... matches) {
            this.matches = new HashSet<>(Arrays.asList(matches));
        }

        static TimeType getType(final String input) {
            TimeType type = null;

            for (final TimeType timeType : values()) {
                if (timeType.matches.contains(input)) {
                    type = timeType;
                    break;
                }
            }

            if (type == null)
                throw new IllegalArgumentException("Failed to find Time Type for String '" + input + "'!");

            return type;
        }

    }

    public class TimeResult {

        private final Long seconds;

        TimeResult(final Long seconds) {
            this.seconds = seconds;
        }

        public Long to(final TimeUnit unit) {
            return unit.convert(seconds, TimeUnit.SECONDS);
        }

    }

}

File: src/main/kotlin/com/github/frcsty/frozenactions/time/TimeAPI.kt
package com.github.frcsty.frozenactions.time

import java.util.concurrent.TimeUnit

private const val DAYS_IN_WEEK = 7
private const val DAYS_IN_MONTH = 30
private const val DAYS_IN_YEAR = 365

/**
 * Parses a String in a "time" format (eg 3m2s) and returns the actual time value in seconds
 */
fun String.parseTime(): TimeResult {
    var seconds = 0L
    val scanner = TimeScanner(
            this
                    .replace(" ", "")
                    .replace("and", "")
                    .replace(",", "")
                    .toLowerCase()
    )
    var next: Long
    while (scanner.hasNext()) {
        next = scanner.nextLong()
        seconds += when (scanner.nextString()) {
            "s", "sec", "secs", "second", "seconds" -> next
            "m", "min", "mins", "minute", "minutes" -> TimeUnit.MINUTES.toSeconds(next)
            "h", "hr", "hrs", "hour", "hours" -> TimeUnit.HOURS.toSeconds(next)
            "d", "dy", "dys", "day", "days" -> TimeUnit.DAYS.toSeconds(next)
            "w", "week", "weeks" -> TimeUnit.DAYS.toSeconds(next * DAYS_IN_WEEK)
            "mo", "mon", "mnth", "month", "months" -> TimeUnit.DAYS.toSeconds(next * DAYS_IN_MONTH)
            "y", "yr", "yrs", "year", "years" -> TimeUnit.DAYS.toSeconds(next * DAYS_IN_YEAR)
            else -> throw IllegalArgumentException()
        }
    }
    return TimeResult(seconds)
}

/**
 * Possibly over-engineering, but this allows us to keep a meaningful return value from [parseTime] while not basically copying [TimeUnit]
 */
@Suppress("EXPERIMENTAL_FEATURE_WARNING")
inline class TimeResult(private val seconds: Long) {

    fun to(unit: TimeUnit): Long {
        return unit.convert(seconds, TimeUnit.SECONDS)
    }
}